Print return values of run tests if type is not boolean.

krk opened the pull request:
If return value of a run test is not boolean, an error is created with
the message "Function return type is not boolean, return value is {}".

- This was very briefly discussed on gitter with @sunfishcode.
- This PR provides a way to introspect into the run test variable values, helpful while writing tests and debugging.

krk committed:
Add print and assertion options to the run tests.

To print the return value of a run test, use "; run print",
to assert on the return value, use "; run %fn() == 42".

abrown commented:
Good idea; I've needed something like this in the past. @nbp and I had talked about extending the `; run` syntax indicating that a function is supposed to be run with something like `run: %fn(3, 5) == 42`. With multiple lines we could then test different arguments to a function and verify their results, presumably then printing out errors like `expected 42, found 99`. Is there any way this could towards that direction?

abrown commented:
For context: https://github.com/bytecodealliance/cranelift/pull/872

krk commented:
I have initially suggested adding a new opcode for printing the argument similar to what `dbg!` does in rust, @sunfishcode suggested that it would be too much work.

If the intention is to print results of comparisons instead of a single value, maybe the function can return a multilane value and we could print the data as two values that are half width of the return type - which would be a workaround.

Extending the run comment `run: %fn(3, 5) == 42` to parse expressions and print both values brings the possibility of inventing a new language that could have more features like other operators, other types of functions etc. I think it would be overly complicated for me to implement at this stage.

On the other hand, one might say that a new `; print "value 1", v1` type of comment command would seem idiomatic w.r.t. the SSA expressions and variable names used in the code. If this is relatively easier to implement than a new `dbg!` opcode, I can try to implement that. Optionally, result of this `;print` command could be available to `check` commands, if that makes sense.

abrown commented:
> I think it would be overly complicated for me to implement at this stage

Makes sense, let's just go with the simple print idea then. I suspect that implementing `; print "value 1", v1` might be just as tricky as adding a new `print` instruction (we would have to mess around inside the function). To keep it simpler and still move towards the comparison idea, could we:
 - keep `; run` with its current logic of only running boolean-returning functions and testing their result (and eventually move this to `; run: %fn(2) > 42`)--optionally we could rename `; run` to `; run: test` for consistency
 - add `; run: print` which would use much of your current code to run a function and then print out the results if it can

What do you think of that?

krk commented:
Would it mean that `; run: print` would make the tests fail always, as it is the case currently?

krk commented:
Sure thing, it makes sense to modify the `run` comment now.

* I have renamed `;run` to `;run: test` and added print and assertion commands.
* Assertion only supports equality and inequality: `; run: %fn() == 42` and `; run: %fn() !=  0x2a`.
* Printing always fails the test and prints the return value: `; run: print`.

I do not like my `parse_options` function, is there an idiomatic way to refactor it into?

abrown commented:
Nice work! I'll take a closer look on Monday.

abrown commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -75,14 +311,41 @@ impl FunctionRunner {
+                    "Function arguments more than supported count: {}",
+                    self.options.args.len()
+                )),
+            },
```
This was a bit hard to read but I think I get it now; it might be more clear with an enum like I mentioned above:
```
match action {
  Print => println!(...),
  Test => check(...),
  TestWithResult => check(...),
}
```

krk replied:
This simplified things a lot!

abrown reviewed (commented):
Thanks for doing this! See my comments for some improvements but I think this approach makes sense. I will use these new features for sure 😄. If anything I said didn't make sense I'll be on the Mozilla IRC if you want to discuss.

bnjbvr reviewed (commented):
Just wanted to add: could we keep the previous behavior that `; run` is equivalent to `; run: test`, so we don't have to modify all the existing test cases, and remove `; run: test` too? (or find a better sub-directive name after the `:`, since "test" is quite overloaded in the context of testing)

krk commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,47 +1,283 @@
+            expected,
+            f32::from_bits(expected as u32),
+            f64::from_bits(expected as u64)
+        ));
```
I used `ConstantData` https://github.com/bytecodealliance/cranelift/blob/5d98701e5cae03160ed1b5a36f5798b3544f43cd/cranelift-codegen/src/ir/constant.rs#L28 as the type for the `expected` value and reused parts of `parse_literals_to_constant_data` in `parser.rs`.

abrown replied:
See my comments about using something like `Value` instead.

krk commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,47 +1,283 @@
+            u128::from_str_radix(&s[2..], 16)
+        } else {
+            u128::from_str_radix(s, 10)
+        }
```
Good catches, I have added tests for supported types in `filetests/function_runner` and added bool and float parsing to `parser::parse_constant_data`.

krk commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,47 +1,283 @@
+        Self {
+            function,
+            isa,
+            options: Self::parse_options(comment),
```
I have introduced a builder `FunctionRunnerBuilder` to unify it with other `with` functions.

krk committed:
Revert verifier errors format string.

abrown commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,13 +179,175 @@ impl FunctionRunner {
+        match s {
+            "==" => Ok(Operator::EQ),
+            "!=" => Ok(Operator::NEQ),
+            _ => Err(format!("Unsupported operator {}", s)),
```
A minor re-factoring suggestion: elsewhere in cranelift I've seen this type of thing as `impl FromStr for Operator { ... }` so that we can just call `.parse()` up on line 320. Not a big deal but it keeps the operator implementation together.

abrown commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,13 +179,175 @@ impl FunctionRunner {
+    /// Build a function runner from an action, a function and the ISA to run on.
+    pub fn finish(self) -> FunctionRunner {
+        FunctionRunner::new(self.function, self.isa.unwrap(), self.action)
+    }
```
Nit: elsewhere in the project we are calling this `build`.

krk replied:
`cranelift-codegen/src/isa/mod.rs` calls it `finish`.

https://github.com/bytecodealliance/cranelift/blob/da996fe5ee4ea32a57156ba93f6813b052964183/cranelift-codegen/src/isa/mod.rs#L151

bnjbvr replied:
Yeah, in this particular case it's this file which is not consistent with the Rust style, so we should fix it (later).

abrown commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,13 +179,175 @@ impl FunctionRunner {
+        match &self.action {
+            FunctionRunnerAction::Test => {
+                if !ret_value_type.is_bool() {
+                    return Err(String::from("Function return type is not boolean"));
```
Let's add the `context.func.name` here as well so that we can more easily find it.

abrown commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,13 +179,175 @@ impl FunctionRunner {
+                if result {
+                    return Ok(());
+                }
+                return Err(format!("Failed: {}", context.func.name.to_string()));
```
Re-factoring nit: I'm not really used to seeing `return` much in Rust 😄:
```
if Self::invoke::<bool>(&code_page) {
  Ok(())
} else {
  Err(format!("Failed: {}", context.func.name.to_string()))
}
```

abrown commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,13 +179,175 @@ impl FunctionRunner {
+                Operator::EQ,
+                ConstantData::default().append(1),
+                true,
+            ),
```
Perhaps we should not overload `check`; why not just `invoke` here and `println!` the returned value?

abrown commented on cranelift-codegen/src/ir/constant.rs:
```diff
@@ -52,6 +57,72 @@ impl From<V128Imm> for ConstantData {
+        Ok(self.into_vec()[0] != 0)
+    }
+}
+
```
I don't think we should modify `ConstantData` for what we are doing; it is designed for constant pool stuff and adding these conversions might make it harder to decouple later from test functionality. Why not build our own `Value` enum:
```
enum Value {
  Bool(bool),
  F32(f32),
  F64(f64),
  ... 
}
```
I wouldn't worry too much about 128-bit types just yet until we get the basics in place. Then I would add a way to parse into this (using Rust's `FromStr/parse`) from text and a type:
```
impl Value {
  fn parse_as(text: &str, type: Type) -> Result<Self, ...> {
    match type {
      B8 | B16 | ... => Ok(Value::Bool(text.parse()?)),
      F32 => Ok(Value::F32(text.parse()?)),
      ...
}
```
Then I would `#[derive(Eq, PartialEq)]` on the enum so we can compare them to each other and add some `From` implementations so that `invoke` can get the raw value (e.g. f32), wrap it in a `Value` with `Value::from` (or `into`), and then compare it against the parsed `Value`s that we get from the comments. (Hopefully that explains the other comments I've made; this isn't the only way to do things but it doesn't involve touching other components like `ConstantData` and it seems natural to Rust--from the little I know about Rust).

abrown replied:
Let me change my tune a bit: I now see down below that you implemented some parsing logic in `parser.rs`; if you want to use that for the `FromStr` on `Value` instead of the built-in Rust parsing that seems OK. We could use methods like `Parser::new(text).match_bool(...)` instead and `Value` may have to use different inner types (e.g. `F32(Ieee32)` instead of `F32(f32)`)--but that seems fine. The underlying thing that I'm getting at is that we likely don't want to touch `ConstantData` in this case.

abrown commented on cranelift-reader/src/parser.rs:
```diff
@@ -864,6 +870,37 @@ impl<'a> Parser<'a> {
+        };
+        Ok(constant_data)
+    }
+
```
See my parsing comments about `Value`; we may not need to change the parser.

abrown commented on docs/testing.rst:
```diff
@@ -397,3 +397,31 @@ Example::
+    }
+    ; run: print
+
+To assert on the return value of the function, add a ``; run: %fn() <op> <expected>`` directive
```
Is it `%fn` or `%trivial_test` here?

abrown replied:
Never mind, looks like it is `%fn` 👍

abrown commented on filetests/function_runner/types.clif:
```diff
@@ -0,0 +1,88 @@
+    v0 = f64const 0x1000.0001
+    return v0
+}
+; run: %fn() == 0x1000.0001
```
Nice work on testing

krk committed:
Implement FromStr for Operator.

krk committed:
Add function name to errors.

krk committed:
Remove some return keywords.

krk committed:
Refactor print_only parameter of check into a new print function.

krk committed:
Add Value enum to represent parsed values in function_runner.

abrown reviewed (approved):
Looks good to me; nice work. We could probably simplify parsing more but we can do that when we finish out the `%fn(42) == 2` syntax (do we have an issue for that, BTW?). @bnjbvr, I'm good to merge but giving you a chance to take a look. In the meantime, @krk, any idea why CI is failing?

abrown commented:
(I will be out next week on vacation and it looks like @bnjbvr may be as well so it might not get merged until after Thanskgiving week).

krk commented:
Only on windows there is a verifier error:

```
2019-11-21T22:50:54.2726499Z FAIL filetests\function_runner\types.clif: run(%test_i128):
2019-11-21T22:50:54.2727654Z Verifier errors
2019-11-21T22:50:54.2727967Z 263 tests
2019-11-21T22:50:54.2728213Z thread 'filetests' panicked at 'test harness: "1 failure"', src\libcore\result.rs:1165:5
2019-11-21T22:50:54.2728470Z stack backtrace:
2019-11-21T22:50:54.2728728Z    0: backtrace::backtrace::trace_unsynchronized
2019-11-21T22:50:54.2728940Z              at C:\Users\VssAdministrator\.cargo\registry\src\github.com-1ecc6299db9ec823\backtrace-0.3.37\src\backtrace\mod.rs:66
2019-11-21T22:50:54.2729193Z    1: std::sys_common::backtrace::_print_fmt
2019-11-21T22:50:54.2729455Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:76
2019-11-21T22:50:54.2729709Z    2: std::sys_common::backtrace::_print::{{impl}}::fmt
2019-11-21T22:50:54.2729958Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:60
2019-11-21T22:50:54.2730196Z    3: core::fmt::write
2019-11-21T22:50:54.2730388Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libcore\fmt\mod.rs:1030
2019-11-21T22:50:54.2730634Z    4: std::io::Write::write_fmt<test::Sink>
2019-11-21T22:50:54.2731057Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\io\mod.rs:1412
2019-11-21T22:50:54.2731568Z    5: std::io::impls::{{impl}}::write_fmt<Write>
2019-11-21T22:50:54.2731827Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\io\impls.rs:141
2019-11-21T22:50:54.2732043Z    6: std::sys_common::backtrace::_print
2019-11-21T22:50:54.2732298Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:64
2019-11-21T22:50:54.2732735Z    7: std::sys_common::backtrace::print
2019-11-21T22:50:54.2732982Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:49
2019-11-21T22:50:54.2733226Z    8: std::panicking::default_hook::{{closure}}
2019-11-21T22:50:54.2733467Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:196
2019-11-21T22:50:54.2733672Z    9: std::panicking::default_hook
2019-11-21T22:50:54.2733917Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:207
2019-11-21T22:50:54.2734390Z   10: std::panicking::rust_panic_with_hook
2019-11-21T22:50:54.2734598Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:473
2019-11-21T22:50:54.2734849Z   11: std::panicking::continue_panic_fmt
2019-11-21T22:50:54.2735087Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:380
2019-11-21T22:50:54.2735324Z   12: std::panicking::rust_begin_panic
2019-11-21T22:50:54.2735566Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:307
2019-11-21T22:50:54.2735767Z   13: core::panicking::panic_fmt
2019-11-21T22:50:54.2736005Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libcore\panicking.rs:85
2019-11-21T22:50:54.2736242Z   14: core::result::unwrap_failed
2019-11-21T22:50:54.2736470Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libcore\result.rs:1165
2019-11-21T22:50:54.2736724Z   15: core::result::Result<core::time::Duration, alloc::string::String>::expect
2019-11-21T22:50:54.2736986Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libcore\result.rs:960
2019-11-21T22:50:54.2737227Z   16: filetests::filetests
2019-11-21T22:50:54.2737452Z              at .\tests\filetests.rs:4
2019-11-21T22:50:54.2737679Z   17: filetests::filetests::{{closure}}
2019-11-21T22:50:54.2737902Z              at .\tests\filetests.rs:2
2019-11-21T22:50:54.2738191Z   18: core::ops::function::FnOnce::call_once<closure-0,()>
2019-11-21T22:50:54.2738481Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libcore\ops\function.rs:227
2019-11-21T22:50:54.2738729Z   19: alloc::boxed::{{impl}}::call_once<(),FnOnce<()>>
2019-11-21T22:50:54.2738972Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\liballoc\boxed.rs:922
2019-11-21T22:50:54.2739206Z   20: panic_unwind::__rust_maybe_catch_panic
2019-11-21T22:50:54.2739443Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libpanic_unwind\lib.rs:80
2019-11-21T22:50:54.2739652Z   21: std::panicking::try
2019-11-21T22:50:54.2739887Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\panicking.rs:271
2019-11-21T22:50:54.2740131Z   22: std::panic::catch_unwind
2019-11-21T22:50:54.2740359Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\panic.rs:394
2019-11-21T22:50:54.2740600Z   23: test::run_test::run_test_inner::{{closure}}
2019-11-21T22:50:54.2741000Z              at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libtest\lib.rs:1413
2019-11-21T22:50:54.2741270Z note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

bnjbvr commented:
I'm happy to give it a look once the Windows failure has been solved. Let us know if you need any help on this!

krk commented:
I think incorrect machine code is generated for Windows x86-64 ABI. 

Return value in the failing test is 128 bits wide and in System V ABI it is expected to be in `rax` and `rdx` whereas Microsoft x64 ABI expects it in `rax` and the stack.

@bnjbvr When `%test_i128` is removed, tests would pass; otherwise I need some pointers to continue :)

Disassembly for `%test_i128` generated on Windows, which is same as it is generated on 64-bit Ubuntu except the calling convention annotation:

```diff
-function %test_i128(i64 fp [%rbp]) -> i64 [%rax], i64 [%rdx], i64 fp [%rbp] fast {
+function %test_i128(i64 fp [%rbp]) -> i64 [%rax], i64 [%rdx], i64 fp [%rbp] system_v {
```

```asm
function %test_i128(i64 fp [%rbp]) -> i64 [%rax], i64 [%rdx], i64 fp [%rbp] fast {
    ss0 = incoming_arg 16, offset -16
 
                                ebb0(v3: i64 [%rbp]):
[RexOp1pushq#50]                    x86_push v3
[RexOp1copysp#8089]                 copy_special %rsp -> %rbp
[RexOp1pu_iq#80b8,%rax]             v0 = iconst.i64 0x4242_3535_1234_5678
[RexOp1pu_iq#80b8,%rcx]             v1 = iconst.i64 0xabcd_ef01_beef_b411
[-,-]                               v2 = iconcat v0, v1
>>>[RexOp1rmov#8089]                   regmove v1, %rcx -> %rdx
[RexOp1popq#58,%rbp]                v4 = x86_pop.i64
[Op1ret#c3]                         return v0, v1, v4
}

Disassembly of 31 bytes:
   0:   40 55                   push    rbp
   2:   48 89 e5                mov     rbp, rsp
   5:   48 b8 78 56 34 12 35 35 42 42
                                movabs  rax, 0x4242353512345678
   f:   48 b9 11 b4 ef be 01 ef cd ab
                                movabs  rcx, 0xabcdef01beefb411
>>>  19:   48 89 ca                mov     rdx, rcx
  1c:   40 5d                   pop     rbp
  1e:   c3                      ret
```

`regmove` is generated in the RA coloring phase:

```rust
DEBUG cranelift_codegen::timing::details      > timing: Starting RA coloring, (during Register allocation)
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring for:
function %test_i128() -> i64 [%rax], i64 [%rdx] system_v {
                                ebb0:
[RexOp1pu_iq#80b8]                  v0 = iconst.i64 0x4242_3535_1234_5678
[RexOp1pu_iq#80b8]                  v1 = iconst.i64 0xabcd_ef01_beef_b411
[-]                                 v2 = iconcat v0, v1
[Op1ret#c3]                         return v0, v1
}

 DEBUG cranelift_codegen::regalloc::coloring   > Coloring ebb0:
 DEBUG cranelift_codegen::regalloc::coloring   > Start ebb0 with entry-diversion set to
      { }
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring v0 = iconst.i64 0x4242_3535_1234_5678
    from [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     color v0 -> %rax
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring v1 = iconst.i64 0xabcd_ef01_beef_b411
    from [ GPR: -cdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     color v1 -> %rcx
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring return v0, v1
    from [ GPR: --db--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::solver     > reassign_in(v0:GPR, %rax -> %rax)
 DEBUG cranelift_codegen::regalloc::solver     > reassign_in(v1:GPR, %rcx -> %rdx)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v0 in %rax (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v1 in %rcx (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::solver     > collect_moves: [v1:GPR(%rcx -> %rdx)]
>>> DEBUG cranelift_codegen::regalloc::solver     > move #0: v1:GPR(%rcx -> %rdx)
 DEBUG cranelift_codegen::timing::details      > timing: Ending RA coloring
```

bnjbvr commented:
Uh, it's an unfortunate conflict with the multi-value returns that emulate what SystemV does, which is incompatible with Win64 here. We can comment out the test in the meanwhile and open an issue about this problem.

Otherwise, it might be needed to investigate deeper in `cranelift-codegen/src/isa/x86/abi.rs` how to pass multiple values with respect to the target ABI convention.

krk committed:
Comment test failing in Windows, issue at #1279.

krk committed:
Fix commented run command in test.

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,301 @@
+
+        macro_rules! consume {
+            ( $ty:ident, $match_fn:expr ) => {{
+                ConstantData::default().append($match_fn)
```
Note: $ty is unused. Could we remove this macro and instead inline it where it's used?

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,301 @@
+            }};
+        }
+
+        fn boolean_to_vec(value: bool) -> Vec<u8> {
```
This function is used once, can we inline it at its single use, and then remove it instead?

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,301 @@
+                Self::I32(ConstantDataU32::try_into(data).unwrap())
+            }
+            types::I64 => {
+                let data = p.match_constant_data(ty).unwrap();
```
It seems we could avoid using ConstantData for types that aren't SIMD, and have something more straightforward avoiding al the ConstantDataU{...} variants.

krk replied:
I have removed ConstantData use for I8, I32 and I64.

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,301 @@
+                    boolean_to_vec(p.match_bool("Expected a boolean").unwrap())
+                );
+                let data = ConstantDataU8::try_into(data).unwrap();
+                Self::Bool(data != 0)
```
Could this just consume the result of `match_bool.unwrap()` instead?

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,301 @@
-        Self::new(function, isa)
+    fn invoke<T>(code_page: &Mmap) -> T
+    where
+        T: std::fmt::Display,
```
It this requirement actually needed here?

krk replied:
No, remnant from debugging, will remove.

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,301 @@
+        if !(func.signature.params.is_empty() && func.signature.returns.len() == 1) {
+            return Err(format!(
+                "Functions must have a signature like: () -> ty, found {}",
+                func.name
```
The error message seems to be incorrect; maybe we can tell which function doesn't respect this, and what is its actual signature?

bnjbvr replied:
Also, it would be worth merging this check in the `return_type` function, and call this function sooner, to avoid compiling if we can't check the result.

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,14 +324,159 @@ impl FunctionRunner {
+        .filter(|r| r.purpose == ArgumentPurpose::Normal)
+        .collect();
+    if normal_rets.len() > 2 || normal_rets.len() == 0 {
+        return Err(String::from("Function return type is not supported"));
```
This should probably read "Number of return values not supported".

bnjbvr commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,14 +324,159 @@ impl FunctionRunner {
+        let both_i64 = normal_rets
+            .into_iter()
+            .map(|p| p.value_type)
+            .all(|ty| ty == types::I64);
```
It seems it would be straightforward to check the type of the two array values at once here.

bnjbvr reviewed (approved):
Looks good to me. There are a few simplifications that could be done, either in this PR or as a follow-up PR. Thanks!

krk committed:
Substitute single line macro "consume".

krk committed:
Remove boolean_to_vec function and simplify its only callsite.

krk committed:
Use ConstantData less in parsing test options.

krk committed:
Remove unnecessary type condition in FunctionRunner.invoke.

krk committed:
Do not compile method that cannot be run in test.

krk committed:
Fix return value count error message.

krk committed:
Simplify return type check for two I64s.

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
+#![allow(trivial_numeric_casts)]
```
Could you please move this to the actual function doing this?

krk replied:
I have removed it instead.

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
 use cranelift_native::builder as host_isa_builder;
+use cranelift_reader::Parser;
+use memmap::Mmap;
 use memmap::MmapMut;
```
`use memmap::{Mmap, MmapMut};`

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
+                    let v = d.into_vec().into_iter().rev();
+                    let mut r: $type = 0;
+                    for b in v {
+                        let (mut shifted, _) = r.overflowing_shl(8);
```
This can't overflow, right?

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
+                Self::I8(data.into())
+            }
+            types::I16 => {
+                let data = p.match_constant_data(ty).unwrap();
```
match_uimm16

krk replied:
I cannot find it, also there is no `Uimm16` type. Should I create them for this PR?

bjorn3 replied:
Makes sense

fitzgen replied:
Was this still going to happen?

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
+            types::I128 => {
+                let data = p.match_constant_data(ty).unwrap();
+                let data = ConstantDataU128::try_into(data).unwrap();
+                Self::I128(data.rotate_left(64))
```
???

krk replied:
Swaps left and right qwords, if causes more problems than it solves, I can remove I128 support from the PR.

bjorn3 replied:
`ConstantData` is stored in little endian format. `ConstantDataU128` converts it to a native endian `u128`, so no swap should be necessary.

fitzgen replied:
It isn't clear to me what the resolution is here, if any.

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
-        let builder = host_isa_builder().expect("Unable to build a TargetIsa for the current host");
-        let isa = builder.finish(flags);
-        Self::new(function, isa)
+    fn invoke<T>(code_page: &Mmap) -> T {
```
unsafe

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
+        if ok {
+            return Ok(());
+        }
+        return Err(format!(
```
Unnecessary `return`.

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,280 @@
+                        if let $t(v) = expected {
+                            Self::check_assertion(Self::invoke::<$z>(&code_page), operator, v)
+                        } else {
+                            panic!()
```
Please use a descriptive message.

bjorn3 commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,15 +303,159 @@ impl FunctionRunner {
+fn return_type(func: &Function) -> Result<types::Type, String> {
+    if !(func.signature.params.is_empty() && func.signature.returns.len() == 1) {
+        return Err(format!(
+            "Functions must have a signature like: () -> ty, found {} for {}",
```
Maybe "function {} must have ..."?

krk committed:
Mark invoke function as unsafe.

krk committed:
Reword error message.

krk committed:
Remove unnecessary return.

krk committed:
Merge use declarations.

krk committed:
Remove overflowing_shl.

krk committed:
Add descriptive panic message.

krk committed:
Remove unnecessary trivial_numeric_casts attribute.

krk commented:
ping @abrown @bnjbvr @bjorn3

bnjbvr commented:
I'm back from vacation, I'll try to take a look at it this week.

fitzgen commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -76,17 +298,161 @@ impl FunctionRunner {
+        let mut parts: Vec<&str> = comment.split_ascii_whitespace().collect();
+
+        if parts.len() < 3 {
+            panic!("Invalid syntax for test assertion, expected \"%fn() <operator> <value>\", found \"{}\"", comment);
```
We shouldn't panic on parse errors, we should return a fallible `ParseResult` like other parsing functions do.

Additionally, it seems a little strange that we have this parsing code in the `cranelift-filetests` crate, when AFAICT all the other parsing code is in the `cranelift-reader` crate. Unless there is a good reason not to, we should match existing conventions and keep the parsing code inside `cranelift-reader`.

fitzgen commented on cranelift-reader/src/parser.rs:
```diff
@@ -609,8 +610,8 @@ impl<'a> Parser<'a> {
-    // Match and consume either a hexadecimal Uimm128 immediate (e.g. 0x000102...) or its literal list form (e.g. [0 1 2...])
-    fn match_constant_data(&mut self, controlling_type: Type) -> ParseResult<ConstantData> {
+    /// Match and consume either a hexadecimal Uimm128 immediate (e.g. 0x000102...) or its literal list form (e.g. [0 1 2...])
+    pub fn match_constant_data(&mut self, controlling_type: Type) -> ParseResult<ConstantData> {
```
I believe that moving that parsing code into this `cranelift-reader` crate would let us avoid making all these methods public as well.

fitzgen commented on cranelift-filetests/src/function_runner.rs:
```diff
@@ -1,53 +1,275 @@
+    ($type: ty, $name: ident) => {
+        struct $name {}
+        impl $name {
+            fn try_into(d: ConstantData) -> Result<$type, String> {
```
Instead of making an ad-hoc `try_into` method, we should follow convention and implement the `TryFrom` trait, which gives us a `try_into` via the blanket impl `TryInto<T> for U` where `U: TryFrom<T>`. For details, see: https://doc.rust-lang.org/stable/std/convert/trait.TryInto.html and https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html

If I'm reading this code right, it would be something like this:

```rust
struct $name(ConstantData);

impl TryFrom<$name> for $type {
    // ...
}
```

But this also makes me wonder why we don't implement `TryFrom<ConstantData> for $type` directly instead of creating all these temporary `ConstantData$Type` types. Is that something that you explored at all? I think it would be the best course to take, unless it doesn't work for some reason or I am overlooking something else.

fitzgen reviewed (commented):
Hi @krk -- thanks for this PR!

I'm not *super* familiar with this area of code, but since Benjamin tagged me for review (I assume because he's got a lot on his plate at the moment) I've taken a look and left a few comments.

Overall, this looks good, but I think there are a few things we could clean up a bit before landing, and it seems like there are still some unresolved discussions from earlier rounds of review that we should figure out before merging this as well.

If any of my comments are unclear, please let me know! I'm happy to clarify.

Cheers!

alexcrichton commented:
Thanks for the PR again, and as a procedural note the Cranelift repository [has now merged](https://github.com/bytecodealliance/cranelift/issues/1408) into the [wasmtime repository](https://github.com/bytecodealliance/wasmtime/tree/master/cranelift).

PRs are no longer landing in this repository, and unfortunately there's no "one button" solution to move a PR to the wasmtime repository. A [script has been prepared](https://gist.github.com/alexcrichton/8cb3f4ef7c25317ba6824ee31e3e53aa), however, to assist you in transferring this PR to the wasmtime repo. Feel free to reach out on Zulip with any questions!

krk commented:
Sorry for abandoning this PR. 

It seems that functionality is being added: https://github.com/bytecodealliance/wasmtime/pull/1223
